README.md
# hdkey

[![NPM Package](https://img.shields.io/npm/v/hdkey.svg?style=flat-square)](https://www.npmjs.org/package/hdkey)
[![build status](https://secure.travis-ci.org/cryptocoinjs/hdkey.svg)](http://travis-ci.org/cryptocoinjs/hdkey)
[![js-standard-style](https://img.shields.io/badge/code%20style-standard-brightgreen.svg)](http://standardjs.com/)

A JavaScript component for
[BIP32](https://github.com/bitcoin/bips/blob/master/bip-0032.mediawiki)(hierarchical
deterministic keys).

## Installation

```bash
npm i --save hdkey
```

## Usage

**example:**

```js
var HDKey = require("hdkey");
var seed =
  "a0c42a9c3ac6abf2ba6a9946ae83af18f51bf1c9fa7dacc4c92513cc4dd015834341c775dcd4c0fac73547c5662d81a9e9361a0aac604a73a321bd9103bce8af";
var hdkey = HDKey.fromMasterSeed(Buffer.from(seed, "hex"));
console.log(hdkey.privateExtendedKey);
// => 'xprv9s21ZrQH143K2SKJK9EYRW3Vsg8tWVHRS54hAJasj1eGsQXeWDHLeuu5hpLHRbeKedDJM4Wj9wHHMmuhPF8dQ3bzyup6R7qmMQ1i1FtzNEW'
console.log(hdkey.publicExtendedKey);
// => 'xpub661MyMwAqRbcEvPmRAmYndzERhyNux1GoHzHxgzVHMBFkCro3kbbCiDZZ5XabZDyXPj5mH3hktvkjhhUdCQxie5e1g4t2GuAWNbPmsSfDp2'
```

### `HDKey.fromMasterSeed(seedBuffer[, versions])`

Creates an `hdkey` object from a master seed buffer. Accepts an optional
`versions` object.

```js
var seed =
  "a0c42a9c3ac6abf2ba6a9946ae83af18f51bf1c9fa7dacc4c92513cc4dd015834341c775dcd4c0fac73547c5662d81a9e9361a0aac604a73a321bd9103bce8af";
var hdkey = HDKey.fromMasterSeed(Buffer.from(seed, "hex"));
```

### `HDKey.fromExtendedKey(extendedKey[, versions, skipVerification])`

Creates an `hdkey` object from a `xprv` or `xpub` extended key string. Accepts
an optional `versions` object & an optional `skipVerification` boolean. If
`skipVerification` is set to true, then the provided public key's x (and y if
uncompressed) coordinate will not will be verified to be on the curve.

```js
var key =
  "xprvA2nrNbFZABcdryreWet9Ea4LvTJcGsqrMzxHx98MMrotbir7yrKCEXw7nadnHM8Dq38EGfSh6dqA9QWTyefMLEcBYJUuekgW4BYPJcr9E7j";
var hdkey = HDKey.fromExtendedKey(key);
```

**or**

```js
var key =
  "xpub6FnCn6nSzZAw5Tw7cgR9bi15UV96gLZhjDstkXXxvCLsUXBGXPdSnLFbdpq8p9HmGsApME5hQTZ3emM2rnY5agb9rXpVGyy3bdW6EEgAtqt";
var hdkey = HDKey.fromExtendedKey(key);
```

### `HDKey.fromJSON(obj)`

Creates an `hdkey` object from an object created via `hdkey.toJSON()`.

---

### `hdkey.derive(path)`

Derives the `hdkey` at `path` from the current `hdkey`.

```js
var seed =
  "fffcf9f6f3f0edeae7e4e1dedbd8d5d2cfccc9c6c3c0bdbab7b4b1aeaba8a5a29f9c999693908d8a8784817e7b7875726f6c696663605d5a5754514e4b484542";
var hdkey = HDKey.fromMasterSeed(Buffer.from(seed, "hex"));
var childkey = hdkey.derive("m/0/2147483647'/1");

console.log(childkey.privateExtendedKey);
// -> "xprv9zFnWC6h2cLgpmSA46vutJzBcfJ8yaJGg8cX1e5StJh45BBciYTRXSd25UEPVuesF9yog62tGAQtHjXajPPdbRCHuWS6T8XA2ECKADdw4Ef"
console.log(childkey.publicExtendedKey);
// -> "xpub6DF8uhdarytz3FWdA8TvFSvvAh8dP3283MY7p2V4SeE2wyWmG5mg5EwVvmdMVCQcoNJxGoWaU9DCWh89LojfZ537wTfunKau47EL2dhHKon"
```

Newer, "hardened" derivation paths look like this:

```js
// as defined by BIP-44
var childkey = hdkey.derive("m/44'/0'/0'/0/0");
```

### `hdkey.sign(hash)`

Signs the buffer `hash` with the private key using `secp256k1` and returns the
signature as a buffer.

### `hdkey.verify(hash, signature)`

Verifies that the `signature` is valid for `hash` and the `hdkey`'s public key
using `secp256k1`. Returns `true` for valid, `false` for invalid. Throws if the
`hash` or `signature` is the wrong length.

### `hdkey.wipePrivateData()`

Wipes all record of the private key from the `hdkey` instance. After calling
this method, the instance will behave as if it was created via
`HDKey.fromExtendedKey(xpub)`.

### `hdkey.toJSON()`

Serializes the `hdkey` to an object that can be `JSON.stringify()`ed.

```js
var seed =
  "fffcf9f6f3f0edeae7e4e1dedbd8d5d2cfccc9c6c3c0bdbab7b4b1aeaba8a5a29f9c999693908d8a8784817e7b7875726f6c696663605d5a5754514e4b484542";
var hdkey = HDKey.fromMasterSeed(Buffer.from(seed, "hex"));

console.log(hdkey.toJSON());
// -> {
//      xpriv: 'xprv9s21ZrQH143K31xYSDQpPDxsXRTUcvj2iNHm5NUtrGiGG5e2DtALGdso3pGz6ssrdK4PFmM8NSpSBHNqPqm55Qn3LqFtT2emdEXVYsCzC2U',
//      xpub: 'xpub661MyMwAqRbcFW31YEwpkMuc5THy2PSt5bDMsktWQcFF8syAmRUapSCGu8ED9W6oDMSgv6Zz8idoc4a6mr8BDzTJY47LJhkJ8UB7WEGuduB'
//    }
```

### `hdkey.privateKey`

Getter/Setter of the `hdkey`'s private key, stored as a buffer.

### `hdkey.publicKey`

Getter/Setter of the `hdkey`'s public key, stored as a buffer.

### `hdkey.privateExtendedKey`

Getter/Setter of the `hdkey`'s `xprv`, stored as a string.

### `hdkey.publicExtendedKey`

Getter/Setter of the `hdkey`'s `xpub`, stored as a string.

## References

- https://github.com/bitcoinjs/bitcoinjs-lib/blob/master/src/hdnode.js
- http://bip32.org/
- http://blog.richardkiss.com/?p=313
- https://github.com/bitcoin/bips/blob/master/bip-0032.mediawiki
- http://bitcoinmagazine.com/8396/deterministic-wallets-advantages-flaw/

## License

MIT


CHANGELOG.md
## 2.1.0 / 2023-01-18

- Add `skipVerification` option to `HDKey.fromExtendedKey()` to allow skipping
  verification logic for performance
  ([#53](https://github.com/cryptocoinjs/hdkey/pull/53))
- Use `ripemd160` package; as Node v18+ `crypto.createHash()` does not support
  ripemd160 ([#51](https://github.com/cryptocoinjs/hdkey/pull/51))
- Fix `.sign()` method in some environments with polyfilled `Buffer`
  implementations ([#50](https://github.com/cryptocoinjs/hdkey/pull/50))
- Performance improvements
  ([#52](https://github.com/cryptocoinjs/hdkey/pull/52))

  2.0.1 / 2020-05-30

---

- Bugfix: prevent mutating buffers passed in
  ([#39](https://github.com/cryptocoinjs/hdkey/pull/39))

  2.0.0 / 2020-05-29

---

- **BREAKING:** Require Node.js v10+
  ([#38](https://github.com/cryptocoinjs/hdkey/pull/38))
- Upgrade `secp256k1` dependency; to use N-API
  ([#32](https://github.com/cryptocoinjs/hdkey/pull/32))

  1.1.2 / 2020-04-16

---

- Fix extremely rare types bug
  ([#33](https://github.com/cryptocoinjs/hdkey/pull/33))
- Use `bs58check` dependency instead of `coinstring`
  ([#30](https://github.com/cryptocoinjs/hdkey/pull/30))
- Don't publish test files
  ([#27](https://github.com/cryptocoinjs/hdkey/issues/27),
  [#34](https://github.com/cryptocoinjs/hdkey/pull/34))

  1.1.1 / 2019-02-09

---

- Fix Electron v4 support. No changes to external API.
  ([#26](https://github.com/cryptocoinjs/hdkey/pull/26))

  1.1.0 / 2018-08-14

---

- Add `wipePrivateData()` method
  ([#22](https://github.com/cryptocoinjs/hdkey/pull/22))
- Add missing LICENSE file
  ([#21](https://github.com/cryptocoinjs/hdkey/pull/21))

  1.0.0 / 2018-05-24

---

- drop support for all Node.js versions 4 and earlier
- fix `derive()` path validation
  ([#20](https://github.com/cryptocoinjs/hdkey/pull/20))

  0.8.0 / 2018-02-06

---

- add `sign()` and `verify()`
- upgrade to `safe-buffer`

  0.7.1 / 2016-05-26

---

- fix bug when `privateKey` is `null`, `privateExtendedKey` should not throw,
  and return `null` [#7][#7]

  0.7.0 / 2016-03-22

---

- upgrade from `ecurve` to `secp256k1`. [#5][#5]

  0.6.0 / 2015-07-02

---

- **breaking** (same day though, haha). Changed
  `publicExtendedKey`/`privateExtendedKey` in `JSON` methods to `xpub`/`xpriv`
- export `HARDENED_OFFSET`

  0.5.0 / 2015-07-02

---

- JavaScript Standard Style
- fix rare condition for BIP32 consistency: [#1][#1]
- added `toJSON()/fromJSON()`

  0.4.0 / 2014-09-24

---

- dropped `sha512` dependency and upgraded to crypto-browserify that supports
  sha512

  0.3.1 / 2014-07-11

---

- removed superfluous code `this._privateKeyBigInteger`

  0.3.0 / 2014-06-29

---

- bugfix: if private key was less than 32 bytes, pad out to 32 bytes with
  leading zeros (this happens in derive)
- changed behavior of `privateExtendedKey()` and `publicExtendedKey()` to return
  base 58 encoded `string` instead of `Buffer`
- changed behavior of `fromExtendedKey()` from accepting a type of `Buffer`
  bytes to base58 `string`

  0.2.0 / 2014-06-25

---

- upgraded `"ecurve": "^0.8.0"` to `"ecurve": "^1.0.0"`
- added functionality to derive public to public child keys

  0.1.0 / 2014-06-16

---

- removed semicolons per http://cryptocoinjs.com/about/contributing/#semicolons
- removed `ECKey` dep
- added `ecurve` dep
- removed `terst` dev dep for `assert`
- added method `fromMasterSeed(seedBuffer, [versions])`
- changed constructor from `new HDKey(masterSeed, [versions])` to
  `new HDKey([versions])`
- added properties: `privateKey` and `publicKey`
- removed method `getIdentifier()`, added property `identifier`
- removed method `getFingerprint()`, added property `fingerprint`
- renamed `private` to `privateExtendedKey`
- renamed `public` to `publicExtendedKey`
- added method `fromExtendedKey()`

  0.0.1 / 2014-05-29

---

- initial release

<!-- generated using jprichardson/issue-links -->

[#7]:
  https://github.com/cryptocoinjs/hdkey/issues/7
  "privateExtendedKey error handling"
[#6]:
  https://github.com/cryptocoinjs/hdkey/pull/6
  "hdkey: use bippath for BIP32 path parsing and validation"
[#5]:
  https://github.com/cryptocoinjs/hdkey/pull/5
  "hdkey: use the secp256k1 package for crypto"
[#4]:
  https://github.com/cryptocoinjs/hdkey/issues/4
  "Is this library still maintained?"
[#3]: https://github.com/cryptocoinjs/hdkey/pull/3 "Update hdkey.js"
[#2]: https://github.com/cryptocoinjs/hdkey/pull/2 "Update hdkey.js"
[#1]:
  https://github.com/cryptocoinjs/hdkey/issues/1
  "rare condition needed for bip consistency"


lib/hdkey.js
/**
 * @typedef HDKey
 * @prop {HDCreate} create
 * @prop {HDFromSeed} fromMasterSeed
 * @prop {HDFromXKey} fromExtendedKey
 * @prop {HDFromJSON} fromJSON
 * @prop {Number} HARDENED_OFFSET - 0x80000000
 */

/**
 * @callback HDCreate
 * @param {HDVersions} [versions]
 * @returns {hdkey}
 */

/**
 * @typedef hdkey
 * @prop {Buffer} chainCode - extra 32-bytes of shared entropy for xkeys
 * @prop {Number} depth - of hd path - typically 0 is seed, 1-3 hardened, 4-5 are not
 * @prop {Buffer} identifier - same bytes as pubKeyHash, but used for id
 * @prop {Number} index - the final segment of an HD Path, the index of the wif/addr
 * @prop {Number} parentFingerprint - 32-bit int, slice of id, stored in child xkeys
 * @prop {Buffer} publicKey
 * @prop {HDVersions} versions - magic bytes for base58 prefix
 * @prop {HDDerivePath} derive - derive a full hd path from the given root
 * @prop {HDDeriveChild} deriveChild - get the next child xkey (in a path segment)
 * @prop {HDFingerprint} getFingerprint
 * @prop {HDMaybeGetString} getPrivateExtendedKey
 * @prop {HDMaybeGetBuffer} getPrivateKey
 * @prop {HDGetString} getPublicExtendedKey
 * @prop {HDSetBuffer} setPublicKey
 * @prop {HDSetBuffer} setPrivateKey
 * @prop {HDSign} sign
 * @prop {HDVerify} verify
 * @prop {HDToJSON} toJSON
 * @prop {HDWipePrivates} wipePrivateData - randomizes private key buffer in-place
 * @prop {Function} _setPublicKey
 */

/** @type {HDKey} */
//@ts-ignore
var HDKey = ("object" === typeof module && exports) || {};
(function (window, HDKey) {
  "use strict";

  let crypto = require("crypto");
  let bs58check = require("bs58check");
  let RIPEMD160 = require("ripemd160");
  let secp256k1 = require("secp256k1");

  let MASTER_SECRET = Buffer.from("Bitcoin seed", "utf8");
  let HARDENED_OFFSET = 0x80000000;
  let LEN = 78;

  // Bitcoin hardcoded by default, can use package `coininfo` for others
  let BITCOIN_VERSIONS = { private: 0x0488ade4, public: 0x0488b21e };

  HDKey.create = function (versions) {
    /** @type {hdkey} */
    let hdkey = {};
    /** @type {Buffer?} */
    let _privateKey = null;

    hdkey.versions = versions || BITCOIN_VERSIONS;
    hdkey.depth = 0;
    hdkey.index = 0;
    //hdkey.publicKey = null;
    //hdkey.identifier = null;
    //hdkey.chainCode = null;
    hdkey.parentFingerprint = 0;

    hdkey.getFingerprint = function () {
      if (!hdkey.identifier) {
        throw new Error("Public key has not been set");
      }
      return hdkey.identifier.slice(0, 4).readUInt32BE(0);
    };

    hdkey.getPrivateKey = function () {
      return _privateKey;
    };
    hdkey.setPrivateKey = function (value) {
      assert(value.length === 32, "Private key must be 32 bytes.");
      assert(secp256k1.privateKeyVerify(value) === true, "Invalid private key");

      _privateKey = value;
      hdkey.publicKey = Buffer.from(secp256k1.publicKeyCreate(value, true));
      hdkey.identifier = hash160(hdkey.publicKey);
    };

    hdkey.setPublicKey = function (value) {
      assert(
        value.length === 33 || value.length === 65,
        "Public key must be 33 or 65 bytes.",
      );
      assert(secp256k1.publicKeyVerify(value) === true, "Invalid public key");
      // force compressed point (performs public key verification)
      let publicKey =
        value.length === 65 ? secp256k1.publicKeyConvert(value, true) : value;
      hdkey._setPublicKey(publicKey);
    };

    /**
     * @param {Buffer} publicKey
     */
    hdkey._setPublicKey = function (publicKey) {
      hdkey.publicKey = Buffer.from(publicKey);
      hdkey.identifier = hash160(publicKey);
      _privateKey = null;
    };

    hdkey.getPrivateExtendedKey = function () {
      if (!_privateKey) {
        return null;
      }

      return bs58check.encode(
        serialize(
          hdkey,
          hdkey.versions.private,
          Buffer.concat([Buffer.alloc(1, 0), _privateKey]),
        ),
      );
    };

    hdkey.getPublicExtendedKey = function () {
      if (!hdkey.publicKey) {
        throw new Error("Missing public key");
      }

      return bs58check.encode(
        serialize(hdkey, hdkey.versions.public, hdkey.publicKey),
      );
    };

    hdkey.derive = function (path) {
      if (path === "m" || path === "M" || path === "m'" || path === "M'") {
        return hdkey;
      }

      let entries = path.split("/");
      let _hdkey = hdkey;
      for (let i = 0; i < entries.length; i += 1) {
        let c = entries[i];
        if (i === 0) {
          assert(/^[mM]{1}/.test(c), 'Path must start with "m" or "M"');
          continue;
        }

        let hardened = c.length > 1 && c[c.length - 1] === "'";
        let childIndex = parseInt(c, 10); // & (HARDENED_OFFSET - 1)
        assert(childIndex < HARDENED_OFFSET, "Invalid index");
        if (hardened) {
          childIndex += HARDENED_OFFSET;
        }

        _hdkey = _hdkey.deriveChild(childIndex);
      }

      return _hdkey;
    };

    hdkey.deriveChild = function (index) {
      let isHardened = index >= HARDENED_OFFSET;
      let indexBuffer = Buffer.allocUnsafe(4);
      indexBuffer.writeUInt32BE(index, 0);

      let data;

      if (isHardened) {
        // Hardened child
        if (!_privateKey) {
          throw new Error("Could not derive hardened child key");
        }

        let pk = _privateKey;
        let zb = Buffer.alloc(1, 0);
        pk = Buffer.concat([zb, pk]);

        // data = 0x00 || ser256(kpar) || ser32(index)
        data = Buffer.concat([pk, indexBuffer]);
      } else {
        // Normal child
        // data = serP(point(kpar)) || ser32(index)
        //      = serP(Kpar) || ser32(index)
        data = Buffer.concat([hdkey.publicKey, indexBuffer]);
      }

      let I = crypto
        .createHmac("sha512", hdkey.chainCode)
        .update(data)
        .digest();
      let IL = I.slice(0, 32);
      let IR = I.slice(32);

      let hd = HDKey.create(hdkey.versions);

      // Private parent key -> private child key
      if (_privateKey) {
        // ki = parse256(IL) + kpar (mod n)
        try {
          hd.setPrivateKey(
            Buffer.from(
              secp256k1.privateKeyTweakAdd(Buffer.from(_privateKey), IL),
            ),
          );
          // throw if IL >= n || (privateKey + IL) === 0
        } catch (err) {
          // In case parse256(IL) >= n or ki == 0, one should proceed with the next value for i
          return hdkey.deriveChild(index + 1);
        }
        // Public parent key -> public child key
      } else {
        // Ki = point(parse256(IL)) + Kpar
        //    = G*IL + Kpar
        try {
          hd.setPublicKey(
            Buffer.from(
              secp256k1.publicKeyTweakAdd(
                Buffer.from(hdkey.publicKey),
                IL,
                true,
              ),
            ),
          );
          // throw if IL >= n || (g**IL + publicKey) is infinity
        } catch (err) {
          // In case parse256(IL) >= n or Ki is the point at infinity, one should proceed with the next value for i
          return hdkey.deriveChild(index + 1);
        }
      }

      hd.chainCode = IR;
      hd.depth = hdkey.depth + 1;
      hd.parentFingerprint = hdkey.getFingerprint();
      hd.index = index;

      return hd;
    };

    hdkey.sign = function (hash) {
      if (!_privateKey) {
        throw new Error("Private Key must be set");
      }

      return Buffer.from(
        secp256k1.ecdsaSign(Uint8Array.from(hash), Uint8Array.from(_privateKey))
          .signature,
      );
    };

    hdkey.verify = function (hash, signature) {
      return secp256k1.ecdsaVerify(
        Uint8Array.from(signature),
        Uint8Array.from(hash),
        Uint8Array.from(hdkey.publicKey),
      );
    };

    hdkey.wipePrivateData = function () {
      if (_privateKey) {
        crypto.randomBytes(_privateKey.length).copy(_privateKey);
      }
      _privateKey = null;
      return hdkey;
    };

    hdkey.toJSON = function () {
      return {
        xpriv: hdkey.getPrivateExtendedKey(),
        xpub: hdkey.getPublicExtendedKey(),
      };
    };

    return hdkey;
  };

  HDKey.fromMasterSeed = function (seedBuffer, versions) {
    let I = crypto
      .createHmac("sha512", MASTER_SECRET)
      .update(seedBuffer)
      .digest();
    let IL = I.slice(0, 32);
    let IR = I.slice(32);

    let hdkey = HDKey.create(versions);
    hdkey.chainCode = IR;
    hdkey.setPrivateKey(IL);

    return hdkey;
  };

  HDKey.fromExtendedKey = function (base58key, versions, skipVerification) {
    // => version(4) || depth(1) || fingerprint(4) || index(4) || chain(32) || key(33)
    versions = versions || BITCOIN_VERSIONS;
    skipVerification = skipVerification || false;
    let hdkey = HDKey.create(versions);

    let keyBuffer = bs58check.decode(base58key);

    let version = keyBuffer.readUInt32BE(0);
    assert(
      version === versions.private || version === versions.public,
      "Version mismatch: does not match private or public",
    );

    hdkey.depth = keyBuffer.readUInt8(4);
    hdkey.parentFingerprint = keyBuffer.readUInt32BE(5);
    hdkey.index = keyBuffer.readUInt32BE(9);
    hdkey.chainCode = keyBuffer.slice(13, 45);

    let key = keyBuffer.slice(45);
    if (key.readUInt8(0) === 0) {
      // private
      assert(
        version === versions.private,
        "Version mismatch: version does not match private",
      );
      hdkey.setPrivateKey(key.slice(1)); // cut off first 0x0 byte
    } else {
      assert(
        version === versions.public,
        "Version mismatch: version does not match public",
      );
      if (skipVerification) {
        hdkey._setPublicKey(key);
      } else {
        hdkey.setPublicKey(key);
      }
    }

    return hdkey;
  };

  HDKey.fromJSON = function (obj) {
    return HDKey.fromExtendedKey(obj.xpriv);
  };

  /**
   * @param {Boolean} assertion
   * @param {String} message
   */
  function assert(assertion, message) {
    if (!assertion) {
      throw new Error(message);
    }
  }

  /**
   * @param {hdkey} hdkey - TODO attach to hdkey
   * @param {Number} version
   * @param {Buffer} key
   */
  function serialize(hdkey, version, key) {
    // => version(4) || depth(1) || fingerprint(4) || index(4) || chain(32) || key(33)
    let buffer = Buffer.allocUnsafe(LEN);

    buffer.writeUInt32BE(version, 0);
    buffer.writeUInt8(hdkey.depth, 4);

    let fingerprint = hdkey.depth ? hdkey.parentFingerprint : 0x00000000;
    buffer.writeUInt32BE(fingerprint, 5);
    buffer.writeUInt32BE(hdkey.index, 9);

    hdkey.chainCode.copy(buffer, 13);
    key.copy(buffer, 45);

    return buffer;
  }

  /**
   * @param {Buffer} buf
   * @returns {Buffer}
   */
  function hash160(buf) {
    let sha = crypto.createHash("sha256").update(buf).digest();
    return new RIPEMD160().update(sha).digest();
  }

  HDKey.HARDENED_OFFSET = HARDENED_OFFSET;
})(("object" === typeof window && window) || {}, HDKey);
if ("object" === typeof module) {
  module.exports = HDKey;
}

// Type Definitions

/**
 * @typedef HDVersions
 * @prop {Number} private - 32-bit int (encodes to 'xprv' in base58)
 * @prop {Number} public - 32-bit int (encodes to 'xpub' in base58)
 */

/**
 * @typedef HDJSON
 * @prop {String?} xpriv - base58check-encoded extended private key
 * @prop {String} xpub - base58check-encoded extended public key
 */

// Function Definitions

/**
 * @callback HDDeriveChild
 * @param {Number} index - includes HARDENED_OFFSET, if applicable
 */

/**
 * @callback HDDerivePath
 * @param {String} path
 */

/**
 * @callback HDFingerprint
 * @returns {Number}
 */

/**
 * @callback HDFromXKey
 * @param {String} base58key - base58check-encoded xkey
 * @param {HDVersions} [versions]
 * @param {Boolean} [skipVerification]
 * returns {hdkey}
 */

/**
 * @callback HDFromJSON
 * @param {HDFromJSONOpts} opts
 * returns {hdkey}
 *
 * @typedef HDFromJSONOpts
 * @prop {String} xpriv
 */

/**
 * @callback HDFromSeed
 * @param {Buffer} seedBuffer
 * @param {HDVersions} [versions]
 */

/**
 * @callback HDGetBuffer
 * @returns {Buffer}
 */

/**
 * @callback HDGetString
 * @returns {String}
 */

/**
 * @callback HDMaybeGetBuffer
 * @returns {Buffer?}
 */

/**
 * @callback HDMaybeGetString
 * @returns {String?}
 */

/**
 * @callback HDSetBuffer
 * @param {Buffer} buf
 */

/**
 * @callback HDSign
 * @param {Buffer} hash
 * @returns {Buffer} - signature
 */

/**
 * @callback HDToJSON
 * @returns {HDJSON}
 */

/**
 * @callback HDVerify
 * @param {Buffer} hash
 * @param {Buffer} signature
 * @returns {Boolean}
 */

/**
 * @callback HDWipePrivates
 */
